### DISCUSSION TITLE
Read Serial Command from WB3S Tuya Chip

### DISCUSSION THREAD

**Discussion by felixwe710** [arendst/Tasmota#10731] [Q&A] [answered]
Hi everyone,

i have a wifi courtain switch and there is an tuya WB3S build in. Later i will replace them with an ESP12f.

Now i have the problem that the WB3S speaks over serial (RX, TX) with another MCU - an MM32F003TW which is controlling the relais.

How can i read the commands which the WB3S sends to the control-MCU so that i know the commands to use the ESP12f later to send them? I am changing the chips to use Tasmota there and connect all to my iobroker.

Here are two images of the controller-board.
The pins are going to the relais etc. but the board on the picture is the item of interest.

![20210128_182958](https://user-images.githubusercontent.com/40203534/106176598-cbacfd80-6197-11eb-8c76-fe8a9293efe8.jpg)
![20210128_182943](https://user-images.githubusercontent.com/40203534/106176631-d9fb1980-6197-11eb-9c7f-06e3d2c46236.jpg)

**Comment by barbudor** [comment, marked as answer]
You need to use you standard USB to serial adapter, the one you would use to serial-flash a ESP module, to listen to the serial connection.
But as you need to listen to both ways you need a little trick like this :
![image](https://user-images.githubusercontent.com/2996042/106181071-6bb95580-619d-11eb-9427-53a3920235a3.png)

As both the WB3S and the TuyaMCU will probably not speak at the same time, this should be ok

On the PC side you need a serial terminal that can display HEX code because standard terminal in ASCII will not be convenient.
On Windows I'm using "Terminal 1.9b by Br@y" that you can download from various places.

**Reply by barbudor** [reply]
Or if you have a 74HCT08 AND gate, that would be even better

![image](https://user-images.githubusercontent.com/2996042/106181404-d7032780-619d-11eb-81b4-d52ad8cd3e0b.png)

**Reply by felixwe710** [reply]
> You need to use you standard USB to serial adapter, the one you would use to serial-flash a ESP module, to listen to the serial connection.
> But as you need to listen to both ways you need a little trick like this :
> ![image](https://user-images.githubusercontent.com/2996042/106181071-6bb95580-619d-11eb-9427-53a3920235a3.png)
> 
> As both the WB3S and the TuyaMCU will probably not speak at the same time, this should be ok
> 
> On the PC side you need a serial terminal that can display HEX code because standard terminal in ASCII will not be convenient.
> On Windows I'm using "Terminal 1.9b by Br@y" that you can download from various places.

Hey barbudor,

thanks for your reply.

I have tryed to read the commands (without a curcuit to read both directions) but the standard settings show me unkown signs.
How should i set up the program from Br@y++ ?

And the next question is: which line are on the left side from the diodes and the HCT08 ?
Are there both TX? 
Should i put the HCT08 at "WB3S to MCU" on the TX line from the WB3S and "MCU to WB3S" on the RX line from the WB3S?

Thanks!

**Reply by barbudor** [reply]
Hi @felixwe710 

Unknown signs can come from either :
- Wrong baud rate
- Trying to display binary messages as ASCII

You need to select the HEX display mode. See on the screenshot below
For the bitrate, you my need to adjust until the HEX code that you see look like Tuya messages starting with 55AA

For the schematic, as written, on the left is your device under test and on the right is your USB serial adapter.
The left signals are
- WB3S TX (to TuyaMCU RX)
- WB3S RX (from TuyaMCU TX)
If you don't have the parts, you can listen to a single signal at a time.

And little mistake from my side, in order to power it with 3V3, it should be a HCT08 but a HC08 (HCT are 5V only).

Sample screen shot of Terminal by Br@y. It's not connected to a TUYA but to my NodeMCU running Tasmota so the baud rate is not what you would need for Tuya. It's just to show what you need to change to get HEX codes and how it looks like:
![image](https://user-images.githubusercontent.com/2996042/106243819-43b40b80-620a-11eb-9a7e-3910da21d97b.png)

**Reply by felixwe710** [reply]
> Hi @felixwe710
> 
> Unknown signs can come from either :
> 
> * Wrong baud rate
> * Trying to display binary messages as ASCII
> 
> You need to select the HEX display mode. See on the screenshot below
> For the bitrate, you my need to adjust until the HEX code that you see look like Tuya messages starting with 55AA
> 
> For the schematic, as written, on the left is your device under test and on the right is your USB serial adapter.
> The left signals are
> 
> * WB3S TX (to TuyaMCU RX)
> * WB3S RX (from TuyaMCU TX)
>   If you don't have the parts, you can listen to a single signal at a time.
> 
> And little mistake from my side, in order to power it with 3V3, it should be a HCT08 but a HC08 (HCT are 5V only).
> 
> Sample screen shot of Terminal by Br@y. It's not connected to a TUYA but to my NodeMCU running Tasmota so the baud rate is not what you would need for Tuya. It's just to show what you need to change to get HEX codes and how it looks like:
> ![image](https://user-images.githubusercontent.com/2996042/106243819-43b40b80-620a-11eb-9a7e-3910da21d97b.png)

Hey,

thank you.

I can ready the HEX Code but how can i use this to send the same command from my ESP?

Can i read the communication in human readable clear text?

Or how does it work?

**Reply by barbudor** [reply]
Have you read https://tasmota.github.io/docs/TuyaMCU/ ?

**Reply by felixwe710** [reply]
Hey,

No, but i'm not so good with these serial communication things.

Normally i use the GPIO's but here i have to use the serial commands...

Can you make me an example how i have to Insert the hex code?

It looks so that you are good in these serial communication things.

Thanks!

**Reply by barbudor** [reply]
Read the chapter, it gives explanations on the Tuya protocol and how to extract the important information
I'm know hardware but not specialist in Tuya
